kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -0,0 +1,128 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.checksum;
+
+import static io.delta.kernel.internal.actions.SingleAction.CHECKPOINT_SCHEMA;
+import static io.delta.kernel.internal.util.Preconditions.checkState;
+
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.exceptions.ChecksumAlreadyExistsException;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.actions.AddFile;
+import io.delta.kernel.internal.actions.DomainMetadata;
+import io.delta.kernel.internal.replay.CreateCheckpointIterator;
+import io.delta.kernel.internal.stats.FileSizeHistogram;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.stream.Collectors;
+
+public class ChecksumUtils {
+  /**
+   * Computes the state of a Delta table and writes a checksum file based on the provided snapshot.
+   *
+   * <p>This method iterates through the table's data using a CreateCheckpointIterator to calculate:
+   *
+   * <ul>
+   *   <li>Total table size in bytes
+   *   <li>Total number of files
+   *   <li>File size histogram
+   *   <li>Domain metadata information
+   * </ul>
+   *
+   * <p>After computing these statistics, it writes a checksum file to the table's log path.
+   *
+   * <p>Note: For very large tables, this operation may be expensive as in worst case, it requires
+   * scanning all log files until the given version.
+   *
+   * @param engine The Engine instance used to access the underlying storage
+   * @param snapshot The SnapshotImpl instance representing the current state of the table
+   * @throws IOException If an I/O error occurs during checksum computation or writing
+   * @throws ChecksumAlreadyExistsException If a checksum already exists for the snapshot version
+   */
+  public static void computeStateAndWriteChecksum(Engine engine, SnapshotImpl snapshot)
+      throws IOException {
+    if (snapshot.getCurrentCrcInfo().isPresent()) {
+      throw new ChecksumAlreadyExistsException(snapshot.getVersion());
+    }
+
+    // TODO: Optimize using last available crc after https://github.com/delta-io/delta/pull/4112
+    LongAdder tableSizeByte = new LongAdder();
+    LongAdder fileCount = new LongAdder();
+    FileSizeHistogram fileSizeHistogram = FileSizeHistogram.createDefaultHistogram();
+    Map<String, DomainMetadata> domainMetadataMap = new HashMap<>();
+    ChecksumWriter checksumWriter = new ChecksumWriter(snapshot.getLogPath());
+
+    // snapshot.getLogSegment() could return last available CRC.
+    // Set minFileRetentionTimestampMillis to infinite future to skip all removed files.
+    try (CreateCheckpointIterator checkpointIterator =
+        new CreateCheckpointIterator(
```

Comment:
we don't need the entire Add or Remove file row. We can optimize this to read just the required column. This can be a separate PR.